# 데모

## 라이브러리 import 및 설정

In [1]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

In [2]:
import gc
from matplotlib import rcParams, pyplot as plt
import numpy as np
import os
import pandas as pd
from pathlib import Path
import re
from sklearn.metrics import accuracy_score, log_loss
from sklearn.model_selection import StratifiedKFold
import tensorflow as tf
from tensorflow.keras import Sequential, layers
from tensorflow.keras.backend import clear_session
from tensorflow.keras.layers import Dense, Embedding, LSTM, GlobalMaxPooling1D, Conv1D, Dropout, Bidirectional
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from tensorflow.keras.utils import plot_model, to_categorical
from tensorflow.keras.optimizers import Adam
import warnings 
warnings.filterwarnings(action='ignore')

In [3]:
rcParams['figure.figsize'] = (16, 8)
plt.style.use('fivethirtyeight')
pd.set_option('max_columns', 100)
pd.set_option("display.precision", 4)
warnings.simplefilter('ignore')

In [4]:
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

## 학습데이터 로드

In [5]:
data_dir = Path('../data')
feature_dir = Path('../feature')
val_dir = Path('../val')
tst_dir = Path('../tst')
sub_dir = Path('../sub')

dirs = [feature_dir, val_dir, tst_dir, sub_dir]
for d in dirs:
    os.makedirs(d, exist_ok=True)

trn_file = data_dir / 'train.csv'
tst_file = data_dir / 'test_x.csv'
sample_file = data_dir / 'sample_submission.csv'

target_col = 'author'
n_fold = 5
n_class = 5
seed = 42

In [6]:
algo_name = 'mta'
feature_name = 'emb'
model_name = f'{algo_name}_{feature_name}'

feature_file = feature_dir / f'{feature_name}.csv'
p_val_file = val_dir / f'{model_name}.val.csv'
p_tst_file = tst_dir / f'{model_name}.tst.csv'
sub_file = sub_dir / f'{model_name}.csv'

In [7]:
train = pd.read_csv(trn_file, index_col=0)
train.head()

text  author
index                                                           
0      He was almost choking. There was so much, so m...       3
1                 “Your sister asked for it, I suppose?”       2
2       She was engaged one day as she walked, in per...       1
3      The captain was in the porch, keeping himself ...       4
4      “Have mercy, gentlemen!” odin flung up his han...       3

In [8]:
test = pd.read_csv(tst_file, index_col=0)
test.head()

text
index                                                   
0      “Not at all. I think she is one of the most ch...
1      "No," replied he, with sudden consciousness, "...
2      As the lady had stated her intention of scream...
3      “And then suddenly in the silence I heard a so...
4      His conviction remained unchanged. So far as I...

## Preprocessing

`MultiHeadSelfAttention` and `TransformerBlock` from the [Text classification with Transformer](https://keras.io/examples/nlp/text_classification_with_transformer/) tutorial at the Keras website.

In [9]:
class MultiHeadSelfAttention(layers.Layer):
    def __init__(self, embed_dim, num_heads=8):
        super(MultiHeadSelfAttention, self).__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        if embed_dim % num_heads != 0:
            raise ValueError(
                f"embedding dimension = {embed_dim} should be divisible by number of heads = {num_heads}"
            )
        self.projection_dim = embed_dim // num_heads
        self.query_dense = layers.Dense(embed_dim)
        self.key_dense = layers.Dense(embed_dim)
        self.value_dense = layers.Dense(embed_dim)
        self.combine_heads = layers.Dense(embed_dim)

    def attention(self, query, key, value):
        score = tf.matmul(query, key, transpose_b=True)
        dim_key = tf.cast(tf.shape(key)[-1], tf.float32)
        scaled_score = score / tf.math.sqrt(dim_key)
        weights = tf.nn.softmax(scaled_score, axis=-1)
        output = tf.matmul(weights, value)
        return output, weights

    def separate_heads(self, x, batch_size):
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.projection_dim))
        return tf.transpose(x, perm=[0, 2, 1, 3])

    def call(self, inputs):
        # x.shape = [batch_size, seq_len, embedding_dim]
        batch_size = tf.shape(inputs)[0]
        query = self.query_dense(inputs)  # (batch_size, seq_len, embed_dim)
        key = self.key_dense(inputs)  # (batch_size, seq_len, embed_dim)
        value = self.value_dense(inputs)  # (batch_size, seq_len, embed_dim)
        query = self.separate_heads(
            query, batch_size
        )  # (batch_size, num_heads, seq_len, projection_dim)
        key = self.separate_heads(
            key, batch_size
        )  # (batch_size, num_heads, seq_len, projection_dim)
        value = self.separate_heads(
            value, batch_size
        )  # (batch_size, num_heads, seq_len, projection_dim)
        attention, weights = self.attention(query, key, value)
        attention = tf.transpose(
            attention, perm=[0, 2, 1, 3]
        )  # (batch_size, seq_len, num_heads, projection_dim)
        concat_attention = tf.reshape(
            attention, (batch_size, -1, self.embed_dim)
        )  # (batch_size, seq_len, embed_dim)
        output = self.combine_heads(
            concat_attention
        )  # (batch_size, seq_len, embed_dim)
        return output
    
    
class TransformerBlock(layers.Layer):
    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
        super(TransformerBlock, self).__init__()
        self.att = MultiHeadSelfAttention(embed_dim, num_heads)
        self.ffn = keras.Sequential(
            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
        )
        self.layernorm1 = layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = layers.LayerNormalization(epsilon=1e-6)
        self.dropout1 = layers.Dropout(rate)
        self.dropout2 = layers.Dropout(rate)

    def call(self, inputs, training):
        attn_output = self.att(inputs)
        attn_output = self.dropout1(attn_output, training=training)
        out1 = self.layernorm1(inputs + attn_output)
        ffn_output = self.ffn(out1)
        ffn_output = self.dropout2(ffn_output, training=training)
        return self.layernorm2(out1 + ffn_output)
    
    
class TokenAndPositionEmbedding(layers.Layer):
    def __init__(self, maxlen, vocab_size, embed_dim):
        super(TokenAndPositionEmbedding, self).__init__()
        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=embed_dim)
        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=embed_dim)

    def call(self, x):
        maxlen = tf.shape(x)[-1]
        positions = tf.range(start=0, limit=maxlen, delta=1)
        positions = self.pos_emb(positions)
        x = self.token_emb(x)
        return x + positions

In [10]:
X_train = train['text'].values
X_test = test['text'].values
y = train['author'].values
print(X_train.shape, X_test.shape, y.shape)

(54879,) (19617,) (54879,)


In [11]:
print(' 최대 길이 : {}'.format(max(len(l) for l in X_train)))
print(' 평균 길이 : {}'.format(sum(map(len, X_train))/len(X_train))) 
#EDA분포에서 어느정도 손실 감안할지 확인하여 최대 길이 500으로 정함

 최대 길이 : 2500
 평균 길이 : 228.1155633302356


In [12]:
contractions = { 
"ain't": "am not / are not / is not / has not / have not",
"aren't": "are not / am not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is / how does",
"I'd": "I had / I would",
"I'd've": "I would have",
"I'll": "I shall / I will",
"I'll've": "I shall have / I will have",
"I'm": "I am",
"I've": "I have",
"isn't": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}

contractions_re = re.compile('(%s)' % '|'.join(contractions.keys()))
def expand_contractions(s, contractions = contractions):
    def replace(match):
        return contractions[match.group(0)]
    return contractions_re.sub(replace, s)

In [33]:
train['text'] = train['text'].apply(expand_contractions)
test['text'] = test['text'].apply(expand_contractions)

train['text'] = train['text'].str.lower()
test['text'] = test['text'].str.lower()

In [ ]:
from autocorrect import Speller 

#create function to spell check strings
def spell_check(x):
    spell = Speller(lang='en')
    return " ".join([spell(i) for i in x.split()])
total['text'] = total['text'].apply(spell_check) ##계산적으로 무거움..시도 X

train['text'] = train['text'].apply(spell_check)
test['text'] = test['text'].apply(spell_check)

## Training

In [34]:
vocab_size = 20000
maxlen = 500
#EDA분포에서 어느정도 손실 감안할지 확인하여 최대 길이 500으로 정함
embed_dim = 64
num_heads = 4  # Number of attention heads
padding_type='post'

In [35]:
tokenizer = Tokenizer(num_words=vocab_size)
tokenizer.fit_on_texts(X_train)
word_index = tokenizer.word_index

In [36]:
train_sequences = tokenizer.texts_to_sequences(X_train)
test_sequences = tokenizer.texts_to_sequences(X_test)

In [37]:
from tensorflow import keras
trn = keras.preprocessing.sequence.pad_sequences(train_sequences, maxlen=maxlen)
tst = keras.preprocessing.sequence.pad_sequences(test_sequences, maxlen=maxlen)
print(trn.shape, tst.shape)

(54879, 500) (19617, 500)


In [38]:
cv = StratifiedKFold(n_splits=n_fold, shuffle=True, random_state=seed)

In [20]:
def get_model():
    ff_dim = 32  # Hidden layer size in feed forward network inside transformer

    inputs = layers.Input(shape=(maxlen,))
    embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
    x = embedding_layer(inputs)
    transformer_block = TransformerBlock(embed_dim, num_heads, ff_dim)
    x = transformer_block(x)
    x = transformer_block(x)
    x = layers.GlobalAveragePooling1D()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Dense(20, activation="relu")(x)
    x = layers.Dropout(0.1)(x)
    outputs = layers.Dense(n_class, activation="softmax")(x)

    model = keras.Model(inputs=inputs, outputs=outputs)
    model.compile(loss='categorical_crossentropy', optimizer=Adam(learning_rate=.001))
    return model


In [21]:
p_val = np.zeros((trn.shape[0], n_class))
p_tst = np.zeros((tst.shape[0], n_class))
for i, (i_trn, i_val) in enumerate(cv.split(trn, y), 1):
    print(f'training model for CV #{i}')
    clf = get_model()
    
    clf.fit(trn[i_trn], 
            to_categorical(y[i_trn]),
            validation_data=(trn[i_val], to_categorical(y[i_val])),
            epochs=10,
            batch_size=64,
            )
    p_val[i_val, :] = clf.predict(trn[i_val])
    p_tst += clf.predict(tst) / n_fold
    
    del clf
    clear_session()
    gc.collect()

training model for CV #1
Epoch 1/12
686/686 [==============================] - 1559s 2s/step - loss: 1.2705 - val_loss: 0.9351
Epoch 2/12
686/686 [==============================] - 1532s 2s/step - loss: 0.7439 - val_loss: 0.7649
Epoch 3/12
686/686 [==============================] - 1528s 2s/step - loss: 0.4974 - val_loss: 0.8015
Epoch 4/12
686/686 [==============================] - 1528s 2s/step - loss: 0.3777 - val_loss: 0.9175
Epoch 5/12
686/686 [==============================] - 1531s 2s/step - loss: 0.3079 - val_loss: 1.0021
Epoch 6/12
686/686 [==============================] - 1529s 2s/step - loss: 0.2604 - val_loss: 1.1193
Epoch 7/12
686/686 [==============================] - 1529s 2s/step - loss: 0.2271 - val_loss: 1.2194
Epoch 8/12
686/686 [==============================] - 1531s 2s/step - loss: 0.2025 - val_loss: 1.2579
Epoch 00008: early stopping
training model for CV #2
Epoch 1/12
686/686 [==============================] - 1531s 2s/step - loss: 1.4267 - val_loss: 1.0971
Epoc

In [22]:
print(f'Accuracy (CV): {accuracy_score(y, np.argmax(p_val, axis=1)) * 100:8.4f}%')
print(f'Log Loss (CV): {log_loss(pd.get_dummies(y), p_val):8.4f}')

Accuracy (CV):  71.7852%
Log Loss (CV):   0.7673


In [ ]:
np.savetxt(p_val_file, p_val, fmt='%.6f', delimiter=',')
np.savetxt(p_tst_file, p_tst, fmt='%.6f', delimiter=',')

## 시각화

In [34]:
print(clf.summary())

Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 250)]        0                                            
__________________________________________________________________________________________________
token_and_position_embedding (T (None, 250, 64)      1296000     input_1[0][0]                    
__________________________________________________________________________________________________
transformer_block (TransformerB (None, 250, 64)      21088       token_and_position_embedding[0][0
                                                                 transformer_block[0][0]          
__________________________________________________________________________________________________
global_average_pooling1d (Globa (None, 64)           0           transformer_block[1][0

In [35]:
#plot_model(clf)

## 제출 파일 생성

In [23]:
sub = pd.read_csv(sample_file, index_col=0)
print(sub.shape)
sub.head()

(19617, 5)


0  1  2  3  4
index               
0      0  0  0  0  0
1      0  0  0  0  0
2      0  0  0  0  0
3      0  0  0  0  0
4      0  0  0  0  0

In [24]:
sub[sub.columns] = p_tst
sub.head()

0       1       2       3       4
index                                        
0      0.0647  0.1183  0.5061  0.2948  0.0161
1      0.0966  0.4303  0.0641  0.1016  0.3074
2      0.9733  0.0112  0.0039  0.0008  0.0107
3      0.0402  0.0094  0.7650  0.0388  0.1466
4      0.3722  0.0912  0.0344  0.4197  0.0825

In [25]:
sub.to_csv(sub_file)

In [26]:
!pip install -U sentencepiece

     |████████████████████████████████| 1.1 MB 13.9 MB/s eta 0:00:01
  Attempting uninstall: sentencepiece
    Found existing installation: sentencepiece 0.1.91
    Uninstalling sentencepiece-0.1.91:
      Successfully uninstalled sentencepiece-0.1.91


In [27]:
!pip install -U tensorflow_hub

In [28]:
!pip install bert-tensorflow

In [39]:
class BahdanauAttention(tf.keras.Model):
  def __init__(self, units):
    super(BahdanauAttention, self).__init__()
    self.W1 = Dense(units)
    self.W2 = Dense(units)
    self.V = Dense(1)

  def call(self, values, query): # 단, key와 value는 같음
    # query shape == (batch_size, hidden size)
    # hidden_with_time_axis shape == (batch_size, 1, hidden size)
    # score 계산을 위해 뒤에서 할 덧셈을 위해서 차원을 변경해줍니다.
    hidden_with_time_axis = tf.expand_dims(query, 1)

    # score shape == (batch_size, max_length, 1)
    # we get 1 at the last axis because we are applying score to self.V
    # the shape of the tensor before applying self.V is (batch_size, max_length, units)
    score = self.V(tf.nn.tanh(
        self.W1(values) + self.W2(hidden_with_time_axis)))

    # attention_weights shape == (batch_size, max_length, 1)
    attention_weights = tf.nn.softmax(score, axis=1)

    # context_vector shape after sum == (batch_size, hidden_size)
    context_vector = attention_weights * values
    context_vector = tf.reduce_sum(context_vector, axis=1)

    return context_vector, attention_weights

In [ ]:
#vocab_size = 20000
#maxlen = 250
maxlen = 500
vocab_size = 20000
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Concatenate, Dropout
from tensorflow.keras import Input, Model
from tensorflow.keras import optimizers
import os

In [40]:
def get_model():
    
    inputs = Input(shape=(maxlen,))
    
    embedding_layer = Embedding(vocab_size, 128, input_length=maxlen, mask_zero = True)(inputs)
    
    lstm = Bidirectional(LSTM(64, dropout=0.5, return_sequences = True))(embedding_layer)
    lstm, forward_h, forward_c, backward_h, backward_c = Bidirectional(LSTM(64, dropout=0.5, |
                                            return_sequences=True, return_state=True))(lstm)
   
    
    state_h = Concatenate()([forward_h, backward_h]) # 은닉 상태
    state_c = Concatenate()([forward_c, backward_c]) # 셀 상태
    
    attention = BahdanauAttention(64) # 가중치 크기 정의
    context_vector, attention_weights = attention(lstm, state_h)
    
    dense1 = Dense(20, activation="relu")(context_vector)

    dropout = Dropout(0.5)(dense1)

    outputs = Dense(n_class, activation="sigmoid")(dropout)

    model = Model(inputs=inputs, outputs=outputs)
    
    model.compile(loss='categorical_crossentropy', optimizer=Adam(learning_rate=.001))
    return model

In [41]:
p_val = np.zeros((trn.shape[0], n_class))
p_tst = np.zeros((tst.shape[0], n_class))
for i, (i_trn, i_val) in enumerate(cv.split(trn, y), 1):
    print(f'training model for CV #{i}')
    clf = get_model()
    
    clf.fit(trn[i_trn], 
            to_categorical(y[i_trn]),
            validation_data=(trn[i_val], to_categorical(y[i_val])),
            epochs=10,
            batch_size=32,
            )
    p_val[i_val, :] = clf.predict(trn[i_val])
    p_tst += clf.predict(tst) / n_fold
    
    del model
    clear_session()
    gc.collect()    

training model for CV #1
Epoch 1/10
1372/1372 [==============================] - 1253s 913ms/step - loss: 1.3603 - val_loss: 0.9928
Epoch 2/10
1372/1372 [==============================] - 1230s 896ms/step - loss: 0.9135 - val_loss: 0.7374
Epoch 3/10
1372/1372 [==============================] - 1222s 891ms/step - loss: 0.6944 - val_loss: 0.6727
Epoch 4/10
1372/1372 [==============================] - 1219s 888ms/step - loss: 0.5693 - val_loss: 0.6960
Epoch 5/10
1372/1372 [==============================] - 1224s 892ms/step - loss: 0.4886 - val_loss: 0.7483
Epoch 6/10
1372/1372 [==============================] - 1223s 891ms/step - loss: 0.4403 - val_loss: 0.7764
Epoch 7/10
1372/1372 [==============================] - 1223s 891ms/step - loss: 0.4017 - val_loss: 0.9171
Epoch 8/10
1372/1372 [==============================] - 1235s 900ms/step - loss: 0.3718 - val_loss: 0.8136
Epoch 9/10
1372/1372 [==============================] - 1242s 905ms/step - loss: 0.3373 - val_loss: 0.8941
Epoch 10/10


In [42]:
print(f'Accuracy (CV): {accuracy_score(y, np.argmax(p_val, axis=1)) * 100:8.4f}%')
print(f'Log Loss (CV): {log_loss(pd.get_dummies(y), p_val):8.4f}')

Accuracy (CV):  75.5116%
Log Loss (CV):   1.0137


In [36]:
print(clf.summary())

Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 250)]        0                                            
__________________________________________________________________________________________________
embedding (Embedding)           (None, 250, 128)     2560000     input_1[0][0]                    
__________________________________________________________________________________________________
bidirectional (Bidirectional)   (None, 250, 128)     98816       embedding[0][0]                  
__________________________________________________________________________________________________
bidirectional_1 (Bidirectional) [(None, 250, 128), ( 98816       bidirectional[0][0]              
_______________________________________________________________________________________

In [43]:
sub = pd.read_csv(sample_file, index_col=0)
sub[sub.columns] = p_tst
sub.to_csv('../sub/AttentionLstmTest.csv')#0.3774 contest

#stopword&alpha_num 여부 X  
 Accuracy (CV):  76.8072%  
 Log Loss (CV):   0.6412...contest   0.33039  
 
#stopword&alpha_num 여부 O  
Accuracy (CV):  72.8238%  
Log Loss (CV):   0.7361  

#stopword&alpha_num 여부 O + earlyStop patience=5-->7에서 멈춤 (이전엔 5였음)  
Accuracy (CV):  72.2808%  
Log Loss (CV):   0.7593  

#stopword&alpha_num 여부 X + maxlen 700  
Accuracy (CV):  71.7852%  
Log Loss (CV):   0.7673.. contest:0.40904


#stopword&alpha_num 여부 X + BahdanauAttention lstm  
Accuracy (CV):  76.1165%  
Log Loss (CV):   0.6881